### PR TITLE
SAK-46436 Add ability for instructors to manually create Tasks that aren't automatically added via tools

### PR DESCRIPTION
**Pull request by ropemar**
In this PR I've added these changes:

**New features**
- You can now add _Tasks widget_ to _Home dashboard_.
- The _Tasks widget_ now shows site related tasks on _Site dashboard_, and all user tasks on _Home dashboard_.
- Instructor/admin users can now create tasks and deliver them to students on a site or groups. Instructors will have a copy of the new task, being able to edit it.
- Students can only edit the description and due dates of their own tasks. This information can't be modified on system-created tasks or tasks delivered to students by instructors.
- The _sakai-editor_ webcomponent can now be initializated with a content attribute.

**Corrections**
- Internationalization of alert message when hard-deleting task was added. It was hardcoded only in english.
- _Sakai-pageable-element_ webcomponent should not reload data when setting _userId_ or _siteId_ properties. This was causing a double invocation that randomized the initialization of data when both properties where defined. This is changed to do data initialization on _connectedCallback_, that is invoked only once at component creation.
- _Sakai-date-picker_ doesn't implement "enable" and "disable" at this time, so It was not working properly. This component is replaced to standard input _datetime-local_.

### CODE REVIEW

**Comment by ropemar** [comment]
@mpellicer you can take a look on this... 👁️👁️

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java
Comment (adrianfish) [inline kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java:43]: You should make this an enum. Have a look at the TopicVisibility and Topic classes in the conversations project.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java:43]: Also, TYPE is a reserved word in Mysql - you should probably use something like ASSIGNED_TYPE

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (adrianfish) [inline kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: Why not use UserTaskAdapterBean here? It has the priority and notes fields contained in it.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (adrianfish) [inline kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:38]: I'd make these one method and just check the list. It seems a bit overkill to have two methods.

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:175]: Not leaving spaces around plus signs is hard to read

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/UserTaskAdapterBean.java
Comment (adrianfish) [inline kernel/api/src/main/java/org/sakaiproject/tasks/api/UserTaskAdapterBean.java:64]: Why not call this "assignedType"?

File: kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java
Comment (adrianfish) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:190]: Please leave these blank lines in. I put them in all my code as I think it makes methods easier to read, less cramped.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java
Comment (adrianfish) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:234]: Please leave a blank line after the method name in these classes. The rest of the code does that as it makes things, imo, easier to read. Only on multiline method bodies though, I don't leave a blank when the method is just one line. That would be crazy.

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:97]: I'd use StringUtils.isNotBlank(bean.getReference())

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:101]: You could just make this one line.

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:117]: I would lift this entire getTitle bit out of the lambda. You are getting tasks for one site. Why do you need to query it again and again?

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:140]:
```suggestion
            log.warn("Error retrieving role on site {} for user {} : {}", siteId, userId, e.toString());
```

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:152]: Why throw your typing away here?

Comment (adrianfish) [reply webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:152]: Grab the site at the top of the method. If you get an IdUnusedException, throw an IllegalArgumentException (siteId must be invalid). Then you can use Collectors.toMap to build your groups map.

**Review by adrianfish — CHANGES_REQUESTED** [review]

File: kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java
Comment (ropemar) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:190]: I personally hate blank lines under curly braces. I think this don't affect readability at all. I've never read java code guidelines that recommend this...

Comment (adrianfish) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:190]: How can you hate a blank line? Please *don't* go and remove them from the rest of the class though.

Comment (adrianfish) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:190]: Intellij IDEA has the option, so blank lines are a thing,  not just my weirdness.

https://intellij-support.jetbrains.com/hc/en-us/community/posts/360000363879-Blank-line-after-function-class-definition

File: kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java
Comment (ropemar) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:234]: Same as above. This is a cosmetic decission, and I don't agree with that.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java
Comment (mpellicer) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java:43]: As Adrian said use enum and " @Enumerated(EnumType.STRING)".

public enum AssignationType {user, site, group}

	@Column(name = "ASSIGNATION_TYPE", length = 5, nullable = false)
        @Enumerated(EnumType.STRING)
	private AssignationType type;

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: The existing method "saveUserTask" was doing more stuff than only saving the info of the database entity "USER_TASKS", since it was saving also the Task. I didn't want to do that, so I implemented a method that only worked with USER_TASKS and the relevant information within this database entity, and named it like it should. I didn't want to re-implement or rename existing code, so I took the decission to overload this method name.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:38]: Overloading methods with different types in a service makes easier to work on the controller side. It's different to assign a task to a single object or to do the same to a collection of them.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java
Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskAssigned.java:43]: Or even just use an ordinal to use less space in the db. That doesn't seem like much, but when you have massive database hosted on Amazon it can mount up.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: Why didn't you want to save the task? It's in the service layer, so multiple db operations are perfectly valid. UserTaskAdapterBean basically has all the information you need to save the task, so adding another method is just confusing.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: Also, don't keep resolving these before people have a chance to look at your comments. When we click the link in the email, we have to go and hunt the resolved block down. It's irritating.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java
Comment (adrianfish) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/tasks/impl/TaskServiceImpl.java:234]: cosmetics are important in code - they can make a big difference to legibility. I think it's easier to read, and the rest of the service has them in since I wrote it. If you really want to make a stand and not have them in your methods then I'll have to accept that, but don't modify existing code, please.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java
Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: I don't want to save the task because what I need is ONLY to save users personal notes/priorities. That's the entity UserTask. Now, some tasks are created by instructors or admin and then they're delivered to students. Students don't have to update tasks, because it would affect to every student on a site or a group. They only can update their own data related to the task, and that's the UserTask entity.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: Ok, so why not just add a boolean flag to not update the Task? I hear what you're saying, but if you think this area of the code needs a rewrite, then let's do that rather than confusing things further.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: I wrote this code, but I can't remember why I save the Task there - I'd have to revisit the code myself. However, I think having this extra method is a code smell. Let's fix the existing piece instead of just rushing it.

Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: You can revisit your code and rename it to some other method name if you want, since it does not save "UserTasks". It does things beyond this simple action. My method does what its name indicates, using primitives in a logic and understandable way.

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: It saves a usertask, what else does it do? Basically, you are being paid to do this work, you have seen an issue in the existing code and are introducing an overloaded method as a way to get around that, instead of fixing it. Why not do the work and fix it, improve what's there already, instead of hacking it?

Comment (adrianfish) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java:26]: When I say you are being paid, I mean you have time and money to do it right, and I don't think this is doing it right.

Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/TaskService.java]: I think you'll be happy with my latest change... 💸 ⌚ 😄

File: webcomponents/tool/src/main/frontend/js/sakai-pageable-element.js
Comment (ropemar) [reply webcomponents/tool/src/main/frontend/js/sakai-pageable-element.js:87]: There was a problem when loading data on the setters when you set both attributes (userId and siteId). It was causing to execute loadData twice. The promise method that was expecting the result of loadData returned their values in random order (sometimes tasks only filtered by userId, sometimes filtered by siteId). Loading data is needed to be retrieved only once, and connectedCallback fits that purpose in the Lit Element lifecycle.

File: webcomponents/tool/src/main/frontend/js/tasks/sakai-tasks-create-task.js
Comment (ropemar) [reply webcomponents/tool/src/main/frontend/js/tasks/sakai-tasks-create-task.js:224]: This was modified because Codacy analysis indicated this had to be within parenthesis.

File: webcomponents/tool/src/main/frontend/js/widgets/sakai-tasks-widget.js
Comment (ropemar) [reply webcomponents/tool/src/main/frontend/js/widgets/sakai-tasks-widget.js:20]: Because the attribute was being converted to the string "undefined" and stored like that in database, causing problems in some spots of the code.

**Comment by mpellicer** [comment]
@ropemar Could you please provide conversion scripts in https://github.com/sakaiproject/sakai-reference?

**Comment by mpellicer** [comment]
Also please see what's going on with the build, I see it broken.

**Comment by ropemar** [comment]
Created sakai-reference PR for conversion scripts: https://github.com/sakaiproject/sakai-reference/pull/153

File: webcomponents/tool/src/main/frontend/js/sakai-pageable-element.js
Comment (adrianfish) [reply webcomponents/tool/src/main/frontend/js/sakai-pageable-element.js:87]: https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks

I understand the issue, but it's the wrong callback. You should either use attributeChangedCallback or check for the other attribute being set, in each of the setters. I prefer using the setters, personally. I used to use attributeChangedCallback but that had the same issue, multiple calls. Read the bit about connectedCallback. It is not guaranteed that the callback fires after attributes being set.

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java:198]: What happens when a user is removed from the group? Does their task get cleaned up?

**Review by adrianfish — CHANGES_REQUESTED** [review]

**Comment by adrianfish** [comment]
I've labelled this as future. We don't want to merge this until 22.0 is out.

File: kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java
Comment (ropemar) [reply kernel/api/src/main/java/org/sakaiproject/tasks/api/Task.java]: Ok, you're right. It is changed now to TASK_OWNER.

**Comment by mpellicer** [comment]
> I've labelled this as future. We don't want to merge this until 22.0 is out.

Could you elaborate please? When you say "We", who are you referring to? As far as I know this piece of work is targeted for 23 which follows a different release plan than Sakai 22 which was frozen many months ago.

**Comment by mpellicer** [comment]
Unless there's more explanation I don't consider this Future.

**Comment by adrianfish** [comment]
Look at the rubrics traverson fix, centralised grading, threaded conversations. They're all future because Earle doesn't want to merge big changes in until we've got 22.0 out. He's the branch manager so I do as he says. It should only be a couple of weeks anyway.

**Comment by mpellicer** [comment]
Okay, works for me.

Thanks @adrianfish for the review and good job @ropemar, I don't see why this is going to cause conflicts in 22.x since the dashboard is almost empty and hidden, maybe I'm wrong, but I'm fine merging and delivering this as is after the 22.0 release, I'm assigning this to my todo list. any new PR related to this work that conflicts with this is going to wait.

@ropemar now is going to work on other community features and there's no plan to work on this feature anymore unless there's bugs reported from the internal QA ,that we're going to do after the merge, or community QA. This piece of work has budget constraints and we need to take care of it since it was done according to the scope and the specs.

Thanks again guys for the good work.

**Comment by adrianfish** [comment]
You've seen my comment about the groups thing though? This can't be merged before that's addressed. This change is creating a user task for each user in the group, but doesn't handle any later mods to the group. Just listening for realm.upd should do it.

**Review by mpellicer — APPROVED** [review]
Good job @ropemar addressing all the requests, as far as I can tell there's no compelling technical concern blocking this work and most of the comments have been addressed, definitely not impacting the feature behavior nor the user experience nor the product roadmap.

**Comment by adrianfish** [comment]
@mpellicer I can see group stuff in the code, so it must be specced. Unless I'm badly mistaken, if you leave this as it is users will be able to see group tasks after they've been removed from a group. If groups are not specced, then remove the group piece entirely.

**Comment by mpellicer** [comment]
> @mpellicer I can see group stuff in the code, so it must be specced. Unless I'm badly mistaken, if you leave this as it is users will be able to see group tasks after they've been removed from a group. If groups are not specced, then remove the group piece entirely.

Yes, it could be, we'll see during QA and decide what's best, probably ended fixing it, but I reckon that if a user reads something and gets deleted from the group, it will not change the fact that the user already read that task, similar to Whatsapp when you read message, if you're kicked from a group you still can read previous content. 

I kinda agree that it's good to purge group tasks in that case, we'll take care of it in a separate Jira.

**Comment by ern** [comment]
Yeah any features with significant changes we want to hold off to after 22.0 is released to avoid merging conflicts.

**Comment by mpellicer** [comment]
Are the dashboard tool and the task list widget included in the 22 release? Have they been well tested by the QA team?

**Comment by ern** [comment]
Yes they have been QA'd

**Comment by adrianfish** [comment]
@mpellicer Why not just fix it before merging it? I mean, what is the point of just creating another JIRA for something you know will be broken and that hasn't even been merged yet? I really don't understand that approach. Just test and fix it now, then merge it. If I *knew* I had some broken stuff in my PR I would never merge it until I'd fixed it.

**Comment by mpellicer** [comment]
I hadn't the chance to test it yet so I can't confirm it's a problem until I test it, bugs are expected after any new development, and the one you're talking about is kinda minor in my opinion. In other words, Roberto will not work more on this PR until we have a list of bugs to fix, maybe he will fix them, maybe other folks from our team.

**Comment by mpellicer** [comment]
> Yes they have been QA'd

Well, I think this is as short as unaccurate, I've reviewed the status for 22 and it can't even tested properly, the report states "Multiple issues still open that need to be fixed before testing can be completed" so far from QA'd in my opinion.

Let's wait for the release and then we'll focus on the new features, thanks for your time guys.

**Comment by mpellicer** [comment]
> @mpellicer Why not just fix it before merging it? I mean, what is the point of just creating another JIRA for something you know will be broken and that hasn't even been merged yet? I really don't understand that approach. Just test and fix it now, then merge it. If I _knew_ I had some broken stuff in my PR I would never merge it until I'd fixed it.

There you go, here's a good example why it makes totally sense.

https://sakaiproject.atlassian.net/browse/SAK-46986

**Comment by adrianfish** [comment]
@mpellicer That was an oversight. I wish someone had have caught that at PR stage, because I don't like putting broken stuff in. You have made my point for me.

**Comment by mpellicer** [comment]
@adrianfish Good, at the end of the day, we take care of what we contribute, Sakai wins.

> @mpellicer That was an oversight. I wish someone had have caught that at PR stage, because I don't like putting broken stuff in. You have made my point for me.

@adrianfish Good, at the end of the day, we take care of what we contribute, Sakai wins.

**Comment by mpellicer** [comment]
@ropemar Could you rebase this PR with the latest changes? We're going to start testing it locally.

Thanks!

**Comment by mpellicer** [comment]
This was tested locally multiple times and was waiting for a long time, but PRs that came later were merged before this and created conflicts, we expect issues with the datepicker.

**Pull request merged**
